make automatic preprocessing optional

QwertyZW opened the pull request:
With this PR callers can chose to opt out of the module's automatic preprocessing by passing the default exported function an optional path to the already preprocessed output.

The API could use a rework, for example allowing the default exported function to take in 2 callbacks instead of 1, for either failure or success.

QwertyZW committed:
make automatic preprocessing optional

taufique71 commented:
Ain't the pre-processing unit already optional? Because if you already have pre-processed code you can directly call lexical unit to tokenize the code text.

``` js
var tokenStream = lexer.lexUnit.tokenize(codeText);
```

QwertyZW commented:
clearPreprocessors clears the preprocessor's output, If I have preprocessed code it would still contain the preprocessor's output in the output file.

So for instance if I have this simple C program

```
#include "stdio.h"

int main(){
    return 0;
}
```

a preprocessor would produce

```
<expanded include statement>

int main(){
    return 0;
}
```

Where as clearPreprocessors produces:

```
int main(){
    return 0;
}
```

It first preprocesses the output, producing a preprocessed file like the former snippet and then clears the preprocessor's output, transforming the output to the latter snippet.

That's why I added a new function with a more explanatory name. This new function is the old clearPreprocessor's main routine. I added a better explanation of what it actually does.

It used to preprocess AND clear the output of the preprocessor.
Now it just clears the output of the preprocessor depending on whether an output file is already defined or not.

With this out of the way, you still need to rethink the API of clearPreprocessors all together.
Consider the following program

```
#include "stdio.h"
int main(){
  #ifdef DBG
  subroutineA();
  #else
  subroutineB();
  #endif
  return 0;
}
```

clearPreprocessors produces

```
int main(){



  subroutineB();

  return 0;
}
```

With no regard to any -D s that are meant to be passed via the commandline.
So in essence  clearPreprocessors doesn't give us an opportunity to do something like this

cpp -Xpreprocessor -DDBG infile.c outfile.i

taufique71 commented:
@QwertyZW Sorry for being late. I understand what your change does. Apart from your suggestion of rethinking the `clearPreprocessor` API can you please provide me a use case of the change you made in the pull request?

QwertyZW commented:
@taufique71  no problem. For instance I was using this on windows, my 'cpp' is installed within the bash environment (Anniversary update). I manually generated the preprocessed c files from within the bash environment and then fed them into the lexer while by passing the lexer's invokation of cpp.

taufique71 commented:
@QwertyZW that is a good use case I guess.

Would you mind editing the README so that people who will be using it can be aware of the feature. Please include the use case while you add it to the README.

Another thing is, as there is no unit tests written yet, it needs to be checked manually. I am a bit busy at the moment so can't do it right now. Can you please ensure that your changes doesn't break any previous features and gives output in the same way as before?

And, thanks for your contribution. :slightly_smiling_face:

QwertyZW committed:
update read me to reflect new feature

QwertyZW committed:
include example cases for new feature in readme

QwertyZW commented:
@taufique71 You're welcome. I was initially looking into developing an independent (within the js context) c indexer for an atom package in js, I ended up at your c-analyzer package which was giving me stacktraces due to the missing command, that's how I ended up tracing that back to here. I think that something like this will eventually require a c preprocessor implemented in js. If you can remove that dependency your 3 packages (lexer, parser and analyzer) will be free of external dependencies other than the node runtime obviously.

As far as the texting goes, I did already test it. I.e the codeText output would be the same as it originally would have been whether you passed the optional arguement or not.
